---

title: Summary of LDtools


keywords: fastai
sidebar: home_sidebar

summary: "Need to be discussed(https://github.com/changebio/LDtools)"
description: "Need to be discussed(https://github.com/changebio/LDtools)"
nb_path: "nbs/02_ldtools.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/02_ldtools.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="From-Gao">From Gao<a class="anchor-link" href="#From-Gao"> </a></h2><h3 id="Objective">Objective<a class="anchor-link" href="#Objective"> </a></h3><p>It is non-trivial to handle all these data merger and region extraction etc for integrative analysis. I would like to revisit and extend LDtools as a utility package to handle summary stats data in the context of data integration -- including fine-mapping, LDSC, PRS, TWAS.</p>
<h3 id="Relevant-existing-tools">Relevant existing tools<a class="anchor-link" href="#Relevant-existing-tools"> </a></h3><p>LDtools developed by @changebio for our analysis pipelines.
Polyfun software has some utility functions to compute LD by chunks, with PLINK input format
<a href="https://github.com/omerwe/polyfun/blob/master/finemapper.py#L476">https://github.com/omerwe/polyfun/blob/master/finemapper.py#L476</a>
as well as with bgen format in which case ldstore is used
<a href="https://github.com/omerwe/polyfun/blob/master/finemapper.py#L385">https://github.com/omerwe/polyfun/blob/master/finemapper.py#L385</a>
bigsnpr::snp_match
<a href="https://rdrr.io/github/MichelNivard/GenomicSEM/man/sumstats.html">https://rdrr.io/github/MichelNivard/GenomicSEM/man/sumstats.html</a>
<a href="https://github.com/precimed/python_convert">https://github.com/precimed/python_convert</a>
<a href="https://github.com/matthijsz/pysumstats">https://github.com/matthijsz/pysumstats</a>
<a href="https://bioconductor.riken.jp/packages/3.14/bioc/vignettes/MungeSumstats/inst/doc/MungeSumstats.html">https://bioconductor.riken.jp/packages/3.14/bioc/vignettes/MungeSumstats/inst/doc/MungeSumstats.html</a>
Google search sumstats or gwas summary statistics there are various repos on github
As a user, I would rather use ldsc code than LDtools because at least they are better tested (by application to multiple papers and dataset). I wish LDtools had used some of these code so we don't reinvent the wheel. However, what polyfun/ldsc has is not good enough.</p>
<p>Also, survey each application fine-mapping, LDSC, PRS, TWAS and comment on how they handle the issue.</p>
<h3 id="Proposal">Proposal<a class="anchor-link" href="#Proposal"> </a></h3><p>Let's take a look at polyfun/ldsc and other related codes, and rework LDtools if their implementation is better. LDtools also contains many other features. Here is the scope of the proposed extended tool:</p>
<p>Support VCF summary statistics format: convert and combine summary stats in VCF format (possibly through rpy2 to the gwas_vcf R package)
Merge summary stats and make sign flips deserves a comprehensive discussion
Compute LD for given region, adjusting for covariates and GRM ; need to support VCF, bgen and plink
Built-in liftover for summary stats or LD index
Other data integration utility features we use repeatedly in xqtl-pipeline, as we move forward with the analysis</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="1.-Working-on">1. Working on<a class="anchor-link" href="#1.-Working-on"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="2.-Issues">2. Issues<a class="anchor-link" href="#2.-Issues"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="3.-Done">3. Done<a class="anchor-link" href="#3.-Done"> </a></h2>
</div>
</div>
</div>
</div>
 

